[RFC] Fix PVS-studio warnings

ZyX-I opened the pull request.

ZyX-I committed:
*: Silence some false positives

ZyX-I commented:
Found a strange code: `goto_buffer()` in `buffer.c` contains a bad code. But if I am not mistaking in Vim it would not be dead because of `do_buffer` located in another position:

```Patch
diff --git a/src/nvim/buffer.c b/src/nvim/buffer.c
index 292eb03..e48b784 100644
--- a/src/nvim/buffer.c
+++ b/src/nvim/buffer.c
@@ -741,12 +741,13 @@ static void clear_wininfo(buf_T *buf)
  */
 void goto_buffer(exarg_T *eap, int start, int dir, int count)
 {
-  (void)do_buffer(*eap->cmd == 's' ? DOBUF_SPLIT : DOBUF_GOTO,
-                  start, dir, count, eap->forceit);
   bufref_T old_curbuf;
   set_bufref(&old_curbuf, curbuf);
   swap_exists_action = SEA_DIALOG;
 
+  (void)do_buffer(*eap->cmd == 's' ? DOBUF_SPLIT : DOBUF_GOTO,
+                  start, dir, count, eap->forceit);
+
   if (swap_exists_action == SEA_QUIT && *eap->cmd == 's') {
     cleanup_T cs;
 
```

ZyX-I committed:
diff: Silence -V519

Not exactly a false positive, but previous assignment is a part of the pattern 
“change global, run code which uses it, change global back”.

ZyX-I committed:
digraph: Ignore false positive

Reversed order is intentional, digraphs allow swapping characters.

ZyX-I committed:
digraph: Fix errors due to has_mbyte and friends being fixed

ZyX-I committed:
edit: Fix strange code

Based on the flow it looks like ptr could not be NULL here: if ptr_arg is NULL 
ptr is compl_leader, if compl_leader is NULL function exits. This also applies 
to Vim as far as I see.

ZyX-I committed:
macros: Fix excessive check

ZyX-I committed:
edit: Copy assert to before the warning

ZyX-I committed:
eval/typval_encode: Silence then/else equivalence warning

ZyX-I committed:
eval: Silence eap->skip false positives

`lnum` starts at `eap->line2` in case of skipping, so cycle is always run at 
least once.

ZyX-I committed:
eval: Fix V547: `d == NULL` was already checked at line 2986

ZyX-I committed:
eval: Refactor f_char2nr

With has_mbyte equal to 1 and &encoding always UTF-8 second argument is no 
longer useful: utf_ptr2char is the same as mb_ptr2char.

Also changes function behaviour a bit: now if second argument is not a number it 
immediately returns with error, without bothering to get a character.

ZyX-I commented:
PVS-Studio is completely wrong at calculating vimvars array length. Don’t know why.

ZyX-I commented:
One of the PVS errors raised a question: may there be a case when partial’s pt_name is NULL and pt_func is NULL?

ZyX-I committed:
eval: Fix useless NULL check

partial_name() as it is written now really cannot return NULL

ZyX-I committed:
eval: Silence octal constant warning

ZyX-I committed:
eval: Refactor nr2char()

Adds error messages, checks type and ignores the second argument.

Currently utf_char2bytes is able to handle any 31-bit character, not
limited by a unicode range. So checking for INT_MAX and not for
something else: function yet uses `int`.

ZyX-I committed:
eval: Fix condition in f_serverstop

ZyX-I committed:
eval: Remove unneeded varp check

ZyX-I committed:
eval: Remove unneeded varp check

ZyX-I committed:
eval: Remove unneeded !eap->skip check

Already checked in the outer if().

ZyX-I committed:
eval: Fix position of buf declaration

ZyX-I committed:
farsi: Simplify condition

ZyX-I committed:
ex_docmd: Remove excessive assignment

ZyX-I committed:
ex_docmd: Remove :Ni! easter egg

ZyX-I committed:
ex_docmd: Remove unneeded if()

ZyX-I committed:
getchar: Fix if block indentation

ZyX-I committed:
hardcopy: Remove unneeded prt_do_conv assignment

ZyX-I committed:
fileio: Simlify help files encoding detection

Most of code is dead when enc_utf8 is always true. Given that `c` is being 
reused for other purposes I left it set to 1 just in case.

ZyX-I commented:
In main.c it complains about possible NULL pointer dereference, pointing at the second of the following lines:

```C
              argv_idx = 0;
              n = get_number_arg(argv[0], &argv_idx, 10);
```

Obviously idx (*second* argument) could not possibly be NULL here, first argument is supposed to be neither. Message:

```
…/src/nvim/main.c|661 col 1| error: V522 Dereferencing of the null pointer 'idx' might take place. The null pointer is passed into 'get_number_arg' function. Inspect the second argument. Check lines: 'main.c:1095'.
```

(note: off-by-two line number in message text, but I loaded that using `:cexpr` which corrected line numbers in the QF list itself)

ZyX-I committed:
memline: Silence “buffer underflow” warning, looks like false positive

ZyX-I committed:
memline: Fix “NULL pointer dereference” warning

It was actually a false positive indicating always-true condition, not real 
dereference.

ZyX-I committed:
message: Remove some enc_utf8/… checks

ZyX-I committed:
message: Some more has_mbyte/enc_utf8 removal

ZyX-I committed:
fileio: Silence “!= identical subexpressions” warning

ZyX-I committed:
os/env: Fix “invalid pointer to local” false positive

ZyX-I committed:
ops: Silence “counter not used in loop” false positive

ZyX-I committed:
normal: Remove unneeded assignment

ZyX-I committed:
normal: Clarify the code

Current variant works only because of PUT_FIXINDENT being equal to true.

ZyX-I committed:
normal: Add figure braces so that code is clearer

ZyX-I committed:
quickfix: Remove unneeded condition

fmt_ptr was checked for being NULL in if() condition earlier.

ZyX-I committed:
quicfix: Remove duplicate condition

ZyX-I committed:
quicfix: Avoid possible NULL dereference

ZyX-I commented:
PVS found suspicious code:

```C
            while (matchpos.lnum == 0
                   && ((options & SEARCH_END) && first_match
                       ?  (nmatched == 1
                           && (int)endpos.col - 1
                           < (int)start_pos.col + extra_col)
                       : ((int)matchpos.col
                          - (ptr[matchpos.col] == NUL)
                          < (int)start_pos.col + extra_col))) {
```

Not sure where placing parens is right. Message:

```
src/nvim/search.c:616 col 1: error: V502 Perhaps the '?:' operator works in a different way than it was expected. The '?:' operator has a lower priority than the '&&' operator.
```

ZyX-I committed:
screen: Remove unneeded check

ZyX-I committed:
screen: Silence “buffer underflow” warning

ZyX-I committed:
screen: Silence NULL dereference false positive

Based on the loop condition when shl_flag is true cur != NULL.

ZyX-I committed:
screen: Remove unneeded condition

Already checked in outer if()

ZyX-I committed:
screen: Remove another portion of has_mbyte/friends-checking stuff

ZyX-I committed:
tag: Silence “buffer underflow” warning

ZyX-I committed:
tag: Fix “initialized twice successively” false positive

ZyX-I committed:
regexp: Fix warning about octal constant

ZyX-I committed:
regexp: Remove another has_mbyte/…-checking stuff

ZyX-I committed:
regexp: Silence octal constant warning

ZyX-I commented:
regexp.c:reg_submatch has two-iteration anti-pattern loop.

ZyX-I committed:
regexp_nfa: Remove another has_mbyte/…-checking code

ZyX-I committed:
regexp_nfa: Remove octal constant

ZyX-I commented:
Warnings left after last commit:

```
src/nvim/buffer.c|750 col 1| warning: V560 A part of conditional expression is always false: swap_exists_action == 2.
src/nvim/buffer.c|826 col 1| warning: V519 The 'swap_exists_action' variable is assigned values twice successively. Perhaps this is a mistake. Check lines: 791, 828.
src/nvim/event/libuv_process.c|47 col 1| warning: V641 The size of the '& proc->in->uv.pipe' buffer is not a multiple of the element size of the type 'uv_stream_t'.
src/nvim/event/libuv_process.c|52 col 1| warning: V641 The size of the '& proc->out->uv.pipe' buffer is not a multiple of the element size of the type 'uv_stream_t'.
src/nvim/event/libuv_process.c|57 col 1| warning: V641 The size of the '& proc->err->uv.pipe' buffer is not a multiple of the element size of the type 'uv_stream_t'.
src/nvim/eval/encode.c|254 col 1| error: V595 The 'state->li->li_tv.vval.v_string' pointer was utilized before it was verified against nullptr. Check lines: 256, 271.
src/nvim/eval.c|1107 col 1| error: V557 Array overrun is possible. The 'prepare_vimvar' function processes value '34'. Inspect the first argument. Check lines: 'eval.c:1145'.
src/nvim/eval.c|1144 col 1| error: V557 Array overrun is possible. The 'VV_VAL' index is pointing beyond array bound.
src/nvim/eval.c|1145 col 1| error: V557 Array overrun is possible. The 'VV_VAL' index is pointing beyond array bound.
src/nvim/eval.c|1120 col 1| error: V557 Array overrun is possible. The 'restore_vimvar' function processes value '34'. Inspect the first argument. Check lines: 'eval.c:1161'.
src/nvim/eval.c|6756 col 1| error: V557 Array overrun is possible. The 'VV_ERRORS' index is pointing beyond array bound.
src/nvim/eval.c|6758 col 1| error: V557 Array overrun is possible. The 'VV_ERRORS' index is pointing beyond array bound.
src/nvim/eval.c|6762 col 1| error: V557 Array overrun is possible. The 'VV_ERRORS' index is pointing beyond array bound.
src/nvim/eval.c|6798 col 1| error: V557 Array overrun is possible. The 'VV_EXCEPTION' index is pointing beyond array bound.
src/nvim/eval.c|6804 col 1| error: V557 Array overrun is possible. The 'VV_EXCEPTION' index is pointing beyond array bound.
src/nvim/eval.c|6807 col 1| error: V557 Array overrun is possible. The 'VV_EXCEPTION' index is pointing beyond array bound.
src/nvim/eval.c|1107 col 1| error: V557 Array overrun is possible. The 'prepare_vimvar' function processes value '34'. Inspect the first argument. Check lines: 'eval.c:8376'.
src/nvim/eval.c|1107 col 1| error: V557 Array overrun is possible. The 'prepare_vimvar' function processes value '35'. Inspect the first argument. Check lines: 'eval.c:8383'.
src/nvim/eval.c|8383 col 1| error: V557 Array overrun is possible. The 'VV_KEY' index is pointing beyond array bound.
src/nvim/eval.c|8399 col 1| error: V557 Array overrun is possible. The 'VV_KEY' index is pointing beyond array bound.
src/nvim/eval.c|8401 col 1| error: V557 Array overrun is possible. The 'VV_KEY' index is pointing beyond array bound.
src/nvim/eval.c|8416 col 1| error: V557 Array overrun is possible. The 'VV_KEY' index is pointing beyond array bound.
src/nvim/eval.c|8424 col 1| error: V557 Array overrun is possible. The 'VV_KEY' index is pointing beyond array bound.
src/nvim/eval.c|1120 col 1| error: V557 Array overrun is possible. The 'restore_vimvar' function processes value '35'. Inspect the first argument. Check lines: 'eval.c:8437'.
src/nvim/eval.c|1120 col 1| error: V557 Array overrun is possible. The 'restore_vimvar' function processes value '34'. Inspect the first argument. Check lines: 'eval.c:8438'.
src/nvim/eval.c|8451 col 1| error: V557 Array overrun is possible. The 'VV_VAL' index is pointing beyond array bound.
src/nvim/eval.c|8452 col 1| error: V557 Array overrun is possible. The 'VV_KEY' index is pointing beyond array bound.
src/nvim/eval.c|8453 col 1| error: V557 Array overrun is possible. The 'VV_VAL' index is pointing beyond array bound.
src/nvim/eval.c|8503 col 1| error: V557 Array overrun is possible. The 'VV_VAL' index is pointing beyond array bound.
src/nvim/eval.c|9319 col 1| error: V557 Array overrun is possible. The 'VV_MOUSE_WIN' index is pointing beyond array bound.
src/nvim/eval.c|9320 col 1| error: V557 Array overrun is possible. The 'VV_MOUSE_WINID' index is pointing beyond array bound.
src/nvim/eval.c|9321 col 1| error: V557 Array overrun is possible. The 'VV_MOUSE_LNUM' index is pointing beyond array bound.
src/nvim/eval.c|9322 col 1| error: V557 Array overrun is possible. The 'VV_MOUSE_COL' index is pointing beyond array bound.
src/nvim/eval.c|9362 col 1| error: V557 Array overrun is possible. The 'VV_MOUSE_WIN' index is pointing beyond array bound.
src/nvim/eval.c|9363 col 1| error: V557 Array overrun is possible. The 'VV_MOUSE_WINID' index is pointing beyond array bound.
src/nvim/eval.c|9364 col 1| error: V557 Array overrun is possible. The 'VV_MOUSE_LNUM' index is pointing beyond array bound.
src/nvim/eval.c|9365 col 1| error: V557 Array overrun is possible. The 'VV_MOUSE_COL' index is pointing beyond array bound.
src/nvim/eval.c|9956 col 1| error: V557 Array overrun is possible. The 'VV_REG' index is pointing beyond array bound.
src/nvim/eval.c|9998 col 1| error: V557 Array overrun is possible. The 'VV_REG' index is pointing beyond array bound.
src/nvim/eval.c|18139 col 1| error: V557 Array overrun is possible. The 'VV_REG' index is pointing beyond array bound.
src/nvim/eval.c|18083 col 1| error: V557 Array overrun is possible. The 'set_vim_var_string' function processes value '31'. Inspect the first argument. Check lines: 'eval.c:18142'.
src/nvim/eval.c|18153 col 1| error: V557 Array overrun is possible. The 'VV_EXCEPTION' index is pointing beyond array bound.
src/nvim/eval.c|18155 col 1| error: V557 Array overrun is possible. The 'VV_EXCEPTION' index is pointing beyond array bound.
src/nvim/eval.c|18168 col 1| error: V557 Array overrun is possible. The 'VV_THROWPOINT' index is pointing beyond array bound.
src/nvim/eval.c|18170 col 1| error: V557 Array overrun is possible. The 'VV_THROWPOINT' index is pointing beyond array bound.
src/nvim/eval.c|21055 col 1| warning: V507 Pointer to local array 'buf' is stored outside the scope of this array. Such a pointer will become invalid.
src/nvim/eval.c|21114 col 1| warning: V614 Potentially uninitialized variable 'wait_start' used. Consider checking the first actual argument of the 'profile_sub_wait' function.
src/nvim/eval.c|18014 col 1| error: V557 Array overrun is possible. The 'get_vim_var_list' function processes value '57'. Inspect the first argument. Check lines: 'eval.c:21799'.
src/nvim/eval.c|18083 col 1| error: V557 Array overrun is possible. The 'set_vim_var_string' function processes value '62'. Inspect the first argument. Check lines: 'eval.c:21843'.
src/nvim/eval.c|18083 col 1| error: V557 Array overrun is possible. The 'set_vim_var_string' function processes value '63'. Inspect the first argument. Check lines: 'eval.c:21844'.
src/nvim/eval.c|18083 col 1| error: V557 Array overrun is possible. The 'set_vim_var_string' function processes value '64'. Inspect the first argument. Check lines: 'eval.c:21845'.
src/nvim/edit.c|3425 col 1| error: V595 The 'ptr' pointer was utilized before it was verified against nullptr. Check lines: 3427, 3436.
src/nvim/event/process.c|82 col 1| warning: V641 The size of the '& proc->in->uv.pipe' buffer is not a multiple of the element size of the type 'uv_stream_t'.
src/nvim/event/process.c|90 col 1| warning: V641 The size of the '& proc->out->uv.pipe' buffer is not a multiple of the element size of the type 'uv_stream_t'.
src/nvim/event/process.c|98 col 1| warning: V641 The size of the '& proc->err->uv.pipe' buffer is not a multiple of the element size of the type 'uv_stream_t'.
src/nvim/event/stream.c|26 col 1| warning: V641 The size of the '& stream' buffer is not a multiple of the element size of the type 'uv_stream_t'.
src/nvim/event/stream.c|52 col 1| warning: V641 The size of the '& stream->uv.pipe' buffer is not a multiple of the element size of the type 'uv_stream_t'.
src/nvim/event/socket.c|74 col 1| warning: V641 The size of the '& watcher->uv.pipe.handle' buffer is not a multiple of the element size of the type 'uv_stream_t'.
src/nvim/event/socket.c|125 col 1| warning: V641 The size of the '& stream->uv.pipe' buffer is not a multiple of the element size of the type 'uv_stream_t'.
src/nvim/main.c|661 col 1| error: V522 Dereferencing of the null pointer 'idx' might take place. The null pointer is passed into 'get_number_arg' function. Inspect the second argument. Check lines: 'main.c:1095'.
src/nvim/memfile.c|108 col 1| error: V547 Expression 'blocksize <= (2147483647 * 2U + 1U)' is always true.
build/src/nvim/auto/options.generated.h|2585 col 1| error: V542 Consider inspecting an odd type cast: 'char' to 'unsigned char *'.
src/nvim/search.c|616 col 1| error: V502 Perhaps the '?:' operator works in a different way than it was expected. The '?:' operator has a lower priority than the '&&' operator.
src/nvim/spell.c|3635 col 1| error: V512 A call of the 'memset' function will lead to underflow of the buffer 'sp'.
src/nvim/shada.c|1278 col 1| warning: V581 The conditional expressions of the 'if' operators situated alongside each other are identical. Check lines: 1276, 1280.
src/nvim/syntax.c|2181 col 1| error: V763 Parameter 'cur_si' is always rewritten in function body before being used.
src/nvim/tui/tui.c|1068 col 1| warning: V641 The size of the '& data->output_handle' buffer is not a multiple of the element size of the type 'uv_stream_t'.
src/nvim/spellfile.c|2469 col 1| warning: V547 Expression 'aff_entry->ae_flags == ((void *) 0)' is always true. Pointer 'aff_entry->ae_flags' == NULL.
src/nvim/regexp.c|6948 col 1| error: V595 The 'retval' pointer was utilized before it was verified against nullptr. Check lines: 6950, 6954.
test/functional/fixtures/tty-test.c|88 col 1| warning: V641 The size of the '& out' buffer is not a multiple of the element size of the type 'uv_stream_t'.
test/functional/fixtures/tty-test.c|149 col 1| warning: V641 The size of the '& tty' buffer is not a multiple of the element size of the type 'uv_stream_t'.
|| 
```

(copied QF buffer).

ZyX-I commented:
I.e. what is left is mostly

1. False positives regarding vimvars wrong length calculation made by PVS-studio.
2. False positives because of casting extended struct to general (e.g. uv_tty_t to uv_stream_t).
3. Some warnings I am not sure what to do with.

ZyX-I committed:
*: Fix linter errors

ZyX-I commented:
Note: this PR is needed to determine whether we need automatic PVS-studio checking. Given the speed it could be done even in neovim/neovim repository, but that needs success condition and no reports could be posted without altering our infrastructure, while neovim/bot-ci is able to generate reports.

The questions:

1. Is this wanted?
2. What do with required comments at the top of the file?
3. What actually is the target? In the current state it looks like we could cope with “no errors from PVS allowed”. At least as long as we ignore LOW certainty level (it adds 218 warnings if I read “Filtered messages: 218” line from plog-converter correctly).

justinmk commented on src/nvim/edit.c:
```diff
@@ -3419,6 +3419,7 @@ static void ins_compl_fixRedoBufForLeader(char_u *ptr_arg)
     else
       return;        /* nothing to do */
   }
+  assert(ptr != NULL);
```
Seems very weird that PVS studio thinks `ptr` can be null here. Could the checks above be optimized out somehow?

ZyX-I replied:
@justinmk If I am not mistaking, PVS studio thinks ptr can be NULL because it sees us checking for it not being NULL.

ZyX-I committed:
buffer: Adjust where do_buffer call is located

It is located there in Vim, but in dd7657c1605246e8f7ade35184069a09dc254e84 
position was for some reason swapped.

ZyX-I commented:
RFC since this may be merged any time now: next commits will only probably fix some problems I did not resolve because I did not understand the code. And in any case I would want to hear answers to https://github.com/neovim/neovim/pull/6533#issuecomment-294369601.

bfredl commented on src/nvim/digraph.c:
```diff
@@ -1700,15 +1697,11 @@ static void printdigraph(digr_T *dp)
+    if (utf_iscomposing(dp->result)) {
+      *p++ = ' ';
     }
+    p += (*mb_char2bytes)(dp->result, p);
```
can remove the `(* )`

bfredl commented on src/nvim/eval.c:
```diff
@@ -7306,18 +7306,14 @@ static void f_changenr(typval_T *argvars, typval_T *rettv, FunPtr fptr)
-    rettv->vval.v_number = (uint8_t)(tv_get_string(&argvars[0])[0]);
   }
+
+  rettv->vval.v_number = utf_ptr2char(
```
nitpick  but I think we should prefer `mb_` over `utf_` whenever possible. Eventually we should only keep the `mb_` variant of functions that now are synonymous.

ZyX-I replied:
In the current state I do not like the names: `mb_ptr2char` is using `utf_ptr2char` which yields one character. `mb_ptr2len` on the other side is using `utfc_ptr2len` which treats composing characters as a part of the character. Both start with `mb_`, but use different definitions of “character” (see the comment near `utfc_ptr2len`, it expcilitly names that cluster a character). I would more like keeping only `utf*` variants then only `mb_*`.

justinmk commented:
> 1. Is this wanted?
> 2. What do with required comments at the top of the file?

Automated PVS would be great. Comments at the top of every file is something that must be maintained (what happens if we have a typo in the comment, that file gets silently ignored?)... Reading https://www.viva64.com/en/b/0457/ , I wonder if some moderate graphic and hyperlink on our README.md and neovim.io would be acceptable instead.

> What actually is the target? In the current state it looks like we could cope with “no errors from PVS allowed”. At least as long as we ignore LOW certainty level (it adds 218 warnings if I read “Filtered messages: 218” line from plog-converter correctly).

That sounds good.

ZyX-I commented:
@justinmk Not silently, it complains to stderr. Not fails the build though.

ZyX-I commented:
Actually, they state that what I am doing in a script is not allowed, including the initial intention of evaluation:

> Note that this mode is not intended to evaluate this software. Please use a demo version or request a temporary license key to try out the analyzer.

ZyX-I commented:
I am personally fine with comments in all C files, would not want to ask anything less. May have clint fail if \*.c file does not start with the comment.

It also appears that comment in question may be modified to fit into 80 columns:

```Patch
diff --git a/scripts/pvscheck.sh b/scripts/pvscheck.sh
index 2e4f990..bfe7692 100755
--- a/scripts/pvscheck.sh
+++ b/scripts/pvscheck.sh
@@ -24,8 +24,8 @@ help() {
 
 get_pvs_comment() {
   cat > pvs-comment << EOF
-// This is an open source non-commercial project. Dear PVS-Studio, please check it.
-// PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
+// This is an open source non-commercial project. Dear PVS-Studio, please check
+// it. PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
 EOF
 }
 
```

justinmk commented on src/nvim/eval.c:
```diff
@@ -12763,25 +12763,32 @@ static void f_nextnonblank(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+    return;
+  }
+  if (num > INT_MAX) {
+    emsgf(_("E5071: Character number could not be greater then INT_MAX (%i)"),
```
"must not" instead of "could not". same for the one above
